lib/pengine/rules.c

```diff
@@ -111,55 +111,56 @@ sort_pairs(gconstpointer a, gconstpointer b, gpointer user_data)
 static void
 populate_hash(xmlNode *nvpair_list, GHashTable *hash, bool overwrite)
 {
-    const char *value = NULL;
-    const char *old_value = NULL;
-    xmlNode *list = nvpair_list;
-    xmlNode *an_attr = NULL;
-
-    if (pcmk__xe_is(list->children, PCMK__XE_ATTRIBUTES)) {
-        list = list->children;
+    if (pcmk__xe_is(nvpair_list->children, PCMK__XE_ATTRIBUTES)) {
+        nvpair_list = nvpair_list->children;
     }
 
-    for (an_attr = pcmk__xe_first_child(list, NULL, NULL, NULL);
+    for (xmlNode *an_attr = pcmk__xe_first_child(nvpair_list, NULL, NULL, NULL);
          an_attr != NULL; an_attr = pcmk__xe_next(an_attr)) {
 
-        if (pcmk__xe_is(an_attr, PCMK_XE_NVPAIR)) {
-            xmlNode *ref_nvpair = pcmk__xe_resolve_idref(an_attr, NULL);
-            const char *name = crm_element_value(ref_nvpair, PCMK_XA_NAME);
+        xmlNode *ref_nvpair = NULL;
+        const char *name = NULL;
+        const char *value = NULL;
+        const char *old_value = NULL;
 
-            value = crm_element_value(an_attr, PCMK_XA_VALUE);
-            if ((value == NULL) && (ref_nvpair != NULL)) {
-                value = crm_element_value(ref_nvpair, PCMK_XA_VALUE);
-            }
+        if (!pcmk__xe_is(an_attr, PCMK_XE_NVPAIR)) {
+            continue;
+        }
 
-            if (name == NULL || value == NULL) {
-                continue;
-            }
+        ref_nvpair = pcmk__xe_resolve_idref(an_attr, NULL);
+        if (ref_nvpair == NULL) {
+            continue;
+        }
 
-            old_value = g_hash_table_lookup(hash, name);
-
-            if (pcmk__str_eq(value, "#default", pcmk__str_casei)) {
-                // @COMPAT Deprecated since 2.1.8
-                pcmk__config_warn("Support for setting meta-attributes (such "
-                                  "as %s) to the explicit value '#default' is "
-                                  "deprecated and will be removed in a future "
-                                  "release", name);
-                if (old_value) {
-                    crm_trace("Letting %s default (removing explicit value \"%s\")",
-                              name, value);
-                    g_hash_table_remove(hash, name);
-                }
-                continue;
-
-            } else if (old_value == NULL) {
-                crm_trace("Setting %s=\"%s\"", name, value);
-                pcmk__insert_dup(hash, name, value);
-
-            } else if (overwrite) {
-                crm_trace("Setting %s=\"%s\" (overwriting old value \"%s\")",
-                          name, value, old_value);
-                pcmk__insert_dup(hash, name, value);
+        name = crm_element_value(ref_nvpair, PCMK_XA_NAME);
+        value = crm_element_value(ref_nvpair, PCMK_XA_VALUE);
+        if ((name == NULL) || (value == NULL)) {
+            continue;
```

Comment:
oddly the schema allows a name without a value. i don't know what would be a use case for that, and ignoring it seems reasonable (the alternative would be dropping any existing value if overwrite). maybe we should make value required in the schema and drop nonconforming entries?

Reply:
Oh yeah I noticed that. It looks like the behavior for 13 years has been to ignore it, so no reason to change that now without good reason. Making it required and dropping value-less nvpairs via XSLT seems reasonable.

Reply:
It's been in the schema that way from the very beginning: 51d1e7b6